cetlvast/suites/docs/examples/example_05_array_memory_resource_alignment.cpp

```diff
@@ -0,0 +1,123 @@
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+///
+
+#include <iostream>
+
+//![example_delegate]
+#include "cetl/pf17/memory_resource.hpp"
+#include "cetl/pmr/array_memory_resource.hpp"
+#include "cetl/pf17/byte.hpp"
+
+namespace cetl
+{
+
+namespace pmr
+{
+
+/// Implementation of cetl::pmr::memory_resource that implements cetl::pf17::pmr::memory_resource using
+/// cetl::pmr::UnsynchronizedArrayMemoryResourceDelegate and cetl::pf17::pmr::memory_resource as the upstream memory
+/// resource.
+class UnsynchronizedArrayMemoryResource final : public cetl::pf17::pmr::memory_resource
+{
+public:
+    using MemoryResourceType = cetl::pf17::pmr::memory_resource;
+
+    /// See cetl::pmr::UnsynchronizedArrayMemoryResourceDelegate for details..
+    UnsynchronizedArrayMemoryResource(void*               buffer,
+                                      std::size_t         buffer_size_bytes,
+                                      MemoryResourceType* upstream = cetl::pf17::pmr::null_memory_resource())
+        : delegate_{buffer, buffer_size_bytes, upstream}
+    {
+    }
+
+    ~UnsynchronizedArrayMemoryResource()                                                   = default;
+    UnsynchronizedArrayMemoryResource(const UnsynchronizedArrayMemoryResource&)            = delete;
+    UnsynchronizedArrayMemoryResource& operator=(const UnsynchronizedArrayMemoryResource&) = delete;
+    UnsynchronizedArrayMemoryResource(UnsynchronizedArrayMemoryResource&&)                 = delete;
+    UnsynchronizedArrayMemoryResource& operator=(UnsynchronizedArrayMemoryResource&&)      = delete;
+
+private:
+    void* do_allocate(std::size_t bytes, std::size_t alignment) override
+    {
+        return delegate_.allocate(bytes, alignment);
+    }
+
+    void do_deallocate(void* p, std::size_t bytes, std::size_t alignment) override
+    {
+        delegate_.deallocate(p, bytes, alignment);
+    }
+
+    bool do_is_equal(const MemoryResourceType& other) const noexcept override
+    {
+        return (this == &other);
+    }
+
+    std::size_t do_max_size() const noexcept override
+    {
+        return delegate_.max_size();
+    }
+
+    void* do_reallocate(void* p, std::size_t old_size_bytes, std::size_t new_size_bytes, std::size_t alignment) override
+    {
+        return delegate_.reallocate(p, old_size_bytes, new_size_bytes, alignment);
+    }
+
+    cetl::pmr::UnsynchronizedArrayMemoryResourceDelegate<MemoryResourceType> delegate_;
+};
+}  // namespace pmr
+}  // namespace cetl
+//![example_delegate]
+
+int main()
+{
+    {
+        static_assert(alignof(std::max_align_t) < 128, "Wow, what hardware are you running on?");
+//![example_0]
+        constexpr std::size_t                        BufferSizeBytes = 64;
+        cetl::pf17::byte                             buffer[BufferSizeBytes];
+        cetl::pmr::UnsynchronizedArrayMemoryResource resource{buffer, BufferSizeBytes};
```

Comment:
Upstream can't be null
```
cetl::pmr::UnsynchronizedArrayMemoryResource resource{buffer, BufferSizeBytes, cetl::pf17::
pmr::null_memory_resource()};
```

Reply:
I'm not following you here. There's a default argument that returns a non-null pointer. Am I not seeing something?

Reply:
Oops, I copy/pasted from my local repo where I had put comments and was out of sync.  Looks like this is fixed now.  Ignore.